### ISSUE TITLE
xrootd-client-4.2.1-1 with XRD_MAKEPATH=true does not emulate old 3.3.6 behavior

### ISSUE THREAD

**Issue by gerardba** [xrootd/xrootd#243]
Hi,

after testing xrdcp from xrootd-client-4.2.1-1 with XRD_MAKEPATH=true I found that it does not reproduce the old 3.3.6 behavior. I logged all the debugging, which I tried against 2 different servers (EOS and dCache) on https://cdcvs.fnal.gov/redmine/issues/9027.

I'll try to summarize below.

My dCache testsuit is:

rpm -q xrootd-client
export XRD_MAKEPATH=true
/usr/bin/xrdcp /etc/passwd root://cmssrmdisk.fnal.gov:1094//dcache/uscmsdisk/store/user/gerard1/
echo $?
/usr/bin/xrdcp /etc/passwd root://cmssrmdisk.fnal.gov:1094//dcache/uscmsdisk/store/user/gerard1/xrdcptest-file-on-exisitng-path
echo $?
/usr/bin/xrdcp /etc/passwd root://cmssrmdisk.fnal.gov:1094//dcache/uscmsdisk/store/user/gerard1/xrdcptest-newpath1/
echo $?
/usr/bin/xrdcp /etc/passwd root://cmssrmdisk.fnal.gov:1094//dcache/uscmsdisk/store/user/gerard1/xrdcptest-newpath2/specificfile-on-new-path
echo $?

And here are the 2 relevant tests:
OLD (3.3.6-4) behaviour:

[root@cmswn1750 ~]# rpm -q xrootd-client
xrootd-client-3.3.6-4.CERN.slc6.x86_64
[root@cmswn1750 ~]# /usr/bin/xrdcp /etc/passwd root://cmssrmdisk.fnal.gov:1094//dcache/uscmsdisk/store/user/gerard1/
xrdcp: Copy to cmssrmdisk.fnal.gov failed on open!
xrdcp: File already exists
[root@cmswn1750 ~]# echo $?
1
[root@cmswn1750 ~]# /usr/bin/xrdcp /etc/passwd root://cmssrmdisk.fnal.gov:1094//dcache/uscmsdisk/store/user/gerard1/xrdcptest-file-on-exisitng-path
[xrootd] Total 0.00 MB    |====================| 100.00 % [inf MB/s]
[root@cmswn1750 ~]# echo $?
0
[root@cmswn1750 ~]# /usr/bin/xrdcp /etc/passwd root://cmssrmdisk.fnal.gov:1094//dcache/uscmsdisk/store/user/gerard1/xrdcptest-newpath1/
[xrootd] Total 0.00 MB    |====================| 100.00 % [0.4 MB/s]
[root@cmswn1750 ~]# echo $?
0
[root@cmswn1750 ~]# /usr/bin/xrdcp /etc/passwd root://cmssrmdisk.fnal.gov:1094//dcache/uscmsdisk/store/user/gerard1/xrdcptest-newpath2/specificfile-on-new-path
[xrootd] Total 0.00 MB    |====================| 100.00 % [0.3 MB/s]
[root@cmswn1750 ~]# echo $?
0
[root@cmswn1750 ~]#

[root@cmschimeradisk ~]# tree /mnt/dcache/uscmsdisk/store/user/gerard1/
/mnt/dcache/uscmsdisk/store/user/gerard1/
├── xrdcptest-file-on-exisitng-path
├── xrdcptest-newpath1
└── xrdcptest-newpath2
    └── specificfile-on-new-path

1 directory, 3 files

New (4.2.1-1) behavior:

[gerard1@dhcp-131-225-86-116 certs]$ rpm -q xrootd-client
xrootd-client-4.2.1-1.osg32.el6.x86_64
[gerard1@dhcp-131-225-86-116 certs]$ export XRD_MAKEPATH=true
[gerard1@dhcp-131-225-86-116 certs]$ /usr/bin/xrdcp /etc/passwd root://cmssrmdisk.fnal.gov:1094//dcache/uscmsdisk/store/user/gerard1/
[1.862kB/1.862kB][100%][==================================================][1.862kB/s]  
[gerard1@dhcp-131-225-86-116 certs]$ echo $?
0
[gerard1@dhcp-131-225-86-116 certs]$ /usr/bin/xrdcp /etc/passwd root://cmssrmdisk.fnal.gov:1094//dcache/uscmsdisk/store/user/gerard1/xrdcptest-file-on-exisitng-path
[1.862kB/1.862kB][100%][==================================================][1.862kB/s]  
[gerard1@dhcp-131-225-86-116 certs]$ echo $?
0
[gerard1@dhcp-131-225-86-116 certs]$ /usr/bin/xrdcp /etc/passwd root://cmssrmdisk.fnal.gov:1094//dcache/uscmsdisk/store/user/gerard1/xrdcptest-newpath1/
[1.862kB/1.862kB][100%][==================================================][1.862kB/s]  
[gerard1@dhcp-131-225-86-116 certs]$ echo $?
0
[gerard1@dhcp-131-225-86-116 certs]$ /usr/bin/xrdcp /etc/passwd root://cmssrmdisk.fnal.gov:1094//dcache/uscmsdisk/store/user/gerard1/xrdcptest-newpath2/specificfile-on-new-path
[1.862kB/1.862kB][100%][==================================================][1.862kB/s]  
[gerard1@dhcp-131-225-86-116 certs]$ echo $?
0
[gerard1@dhcp-131-225-86-116 certs]$

#Everything works, let's see what it did...
[root@cmschimeradisk ~]# tree /mnt/dcache/uscmsdisk/store/user/gerard1/
/mnt/dcache/uscmsdisk/store/user/gerard1/
├── passwd
├── xrdcptest-file-on-exisitng-path
│   └── passwd
├── xrdcptest-newpath1
│   └── passwd
└── xrdcptest-newpath2
    └── specificfile-on-new-path
        └── passwd

4 directories, 4 files

The final path/file names are different, so I believe this will not fix our issue. Do you think the old behaviour could be emulated?

thanks,
  Gerard

**Comment by abh3** [comment]
Hi Gerard,

Indeed, it does not and this is a sticky situation because emulatng the 
old behaviour is not correct either.

In the case:

3.3.6
/usr/bin/xrdcp /etc/passwd
root://cmssrmdisk.fnal.gov:1094//dcache/uscmsdisk/store/user/gerard1/
xrdcp: Copy to cmssrmdisk.fnal.gov failed on open!
xrdcp: File already exists

4.4.2
/usr/bin/xrdcp /etc/passwd
root://cmssrmdisk.fnal.gov:1094//dcache/uscmsdisk/store/user/gerard1/

Generally, a specification like the above should copy the source file 
into the specified destination dirctory, if it exists. Arugably, 4.4.2 
produced the correct result and 3.3.6 did not.

The question then becomes what should happen if the final component in the 
target specification does not exist. In 3.3.6 the final component is 
always treated as a destination file name specification while 4.4.2 treats 
it as a directory specification. I would agree that is incorrect. It 
would appear that there is a conflict between path creation and file
creation. So, this is not exactly an simple fix.

The bottom line is we can probably make it close to 3.3.6 behaviour but 
making it identical would not be the way to go.

Anyway, I will see what we can do here to make xrdcp be in line with other 
copy programs like scp and cp. Though, that isn't a slam dunk since 
neither have a way to automatically create a path if the source is a 
regular file.

Andy

**Comment by simonmichal** [comment]
Hi guys,
I did some quick tests with 4.2.1 (both client and server) and that's the behaviour that I observed

on the server:
[simonm@slc6-test ~]$ ls -ltr /tmp/
total 0
(the /tmp is empty)

on the client:
[simonm@idefix test]$ xrdcp ./test.txt root://slc6-test:1094//tmp
[10B/10B][100%][==================================================][10B/s]
(copying a file works)

[0B/0B][100%][==================================================][0B/s]  
Run: [ERROR] Server responded with an error: [3006] Unable to create /tmp/test.txt; file exists
(copying the file if the file already exists fails which I suppose is the right behaviour)

[simonm@idefix test]$ xrdcp -f ./test.txt root://slc6-test:1094//tmp
[10B/10B][100%][==================================================][10B/s]
(copying the file if the file exists succeeded with the force flag, again correct behaviour)

[simonm@idefix test]$ xrdcp ./test.txt root://slc6-test:1094//tmp/new_name
[10B/10B][100%][==================================================][10B/s]
(copying the file with new name: new_name)

on the server:
[simonm@slc6-test ~]$ ls -ltr /tmp/
total 8
-rw-r--r--. 1 xrootd xrootd 10 Jun 19 17:08 test.txt
-rw-r--r--. 1 xrootd xrootd 10 Jun 19 17:09 new_name
(both files are there, every thing seems to be fine)

on the client:
[simonm@idefix test]$ echo $XRD_MAKEPATH 
true
[simonm@idefix test]$ xrdcp ./test.txt root://slc6-test:1094//tmp/new_dir/
[0B/0B][100%][==================================================][0B/s]  
Run: [ERROR] Server responded with an error: [3016] Unable to create /tmp/new_dir/; is a directory
(it fails, complains that new_dir is a directory, although the XRD_MAKEPATH is set to true)

on the server:
[simonm@slc6-test ~]$ ls -ltr /tmp/
total 12
-rw-r--r--. 1 xrootd xrootd   10 Jun 19 17:08 test.txt
-rw-r--r--. 1 xrootd xrootd   10 Jun 19 17:09 new_name
drwxr-xr-x. 2 xrootd xrootd 4096 Jun 19 17:11 new_dir
(the 'new_dir' has been created)
[simonm@slc6-test ~]$ ls -ltr /tmp/new_dir/
total 0
(but it's empty)

on the client:
[simonm@idefix test]$ xrdcp ./test.txt root://slc6-test:1094//tmp/new_dir/
[10B/10B][100%][==================================================][10B/s]
(this time it works because the directory is already there)

on the server:
[simonm@slc6-test ~]$ ls -ltr /tmp/new_dir/
total 4
-rw-r--r--. 1 xrootd xrootd 10 Jun 19 17:15 test.txt

on the client:
[simonm@idefix test]$ xrdcp ./test.txt root://slc6-test:1094//tmp/other_dir/new_name
[10B/10B][100%][==================================================][10B/s]
(copy the file to new directory with new name, it worked!)

on the server:
[simonm@slc6-test ~]$ ls -ltr /tmp/
total 16
-rw-r--r--. 1 xrootd xrootd   10 Jun 19 17:08 test.txt
-rw-r--r--. 1 xrootd xrootd   10 Jun 19 17:09 new_name
drwxr-xr-x. 2 xrootd xrootd 4096 Jun 19 17:15 new_dir
drwxr-xr-x. 2 xrootd xrootd 4096 Jun 19 17:16 other_dir

[simonm@slc6-test ~]$ ls -ltr /tmp/other_dir/
total 4
-rw-r--r--. 1 xrootd xrootd 10 Jun 19 17:16 new_name
(so now it worked, xrootd managed to create the new directory and copy the file with new name)


To summarize: the only problem that I observed is that if a file is being copied to non-existent directory it fails (the directory is created but the file is not copied). Surprisingly if a file is being copied to non-existent directory with a different name everything is fine.

Michal

**Comment by abh3** [comment]
Michal is correct. I have done exhaustive tests which revealed that the differences are due to the presence of a trailing slash in the destination specification. A simple test using /tmp/x44 as the destination follows:

***Last component does not exist and dest has trailing slash (results differ)
3.3.6
./xrdcp /etc/passwd root://rhel6-64a//tmp/x44/newdir/
[xrootd] Total 0.00 MB  |====================| 100.00 % [0.3 MB/s]
ls -lr /tmp/x44
total 4
-rw-r--r-- 1 abh sf 2018 Jun 20 15:02 newdir

4.2.1
./xrdcp /etc/passwd root://rhel6-64a//tmp/x44/newdir/
[0B/0B][100%][==================================================][0B/s]  
Run: [ERROR] Server responded with an error: [3016] Unable to create /tmp/x44/newdir/; is a directory
sls -lr /tmp/x44
total 4
drwxr-xr-x 2 abh sf 4096 Jun 20 14:58 newdir/

***Last component does not exist and dest does not have a trailing slash (results identical)
3.3.6
./xrdcp /etc/passwd root://rhel6-64a//tmp/x44/newdir
[xrootd] Total 0.00 MB  |====================| 100.00 % [0.3 MB/s]
ls -lr /tmp/x44
total 4
-rw-r--r-- 1 abh sf 2018 Jun 20 15:04 newdir

4.2.1
./xrdcp /etc/passwd root://rhel6-64a//tmp/x44/newdir
[1.971kB/1.971kB][100%][==================================================][1.971kB/s]
ls -lr /tmp/x44
total 4
-rw-r--r-- 1 abh sf 2018 Jun 20 14:54 newdir

*** Last component exists and is a directory and dest has trailing slash (results differ)
3.3.6
/xrdcp /etc/passwd root://rhel6-64a//tmp/x44/
xrdcp: Copy to rhel6-64a.slac.stanford.edu failed on open!
xrdcp: Unable to create /tmp/x44; file exists
ls -lr /tmp/x44
total 0

4.2.1
./xrdcp /etc/passwd root://rhel6-64a//tmp/x44/
[1.971kB/1.971kB][100%][==================================================][1.971kB/s]
ls -lr /tmp/x44
total 4
-rw-r--r-- 1 abh sf 2018 Jun 20 14:59 passwd

*** Last component exists and is a directory and dest does not have a trailing slash (results differ)
3.3.6
./xrdcp /etc/passwd root://rhel6-64a//tmp/x44
xrdcp: Copy to rhel6-64a.slac.stanford.edu failed on open!
xrdcp: Unable to create /tmp/x44; file exists
ls -lr /tmp/x44
total 4
-rw-r--r-- 1 abh sf 2018 Jun 20 14:59 passwd

4.2.1
./xrdcp /etc/passwd root://rhel6-64a//tmp/x44
[1.971kB/1.971kB][100%][==================================================][1.971kB/s]
ls -lr /tmp/x44
total 4
-rw-r--r-- 1 abh sf 2018 Jun 20 15:00 passwd

--------------------

I will discuss these results in a subsequent post.

**Comment by abh3** [comment]
The difference between 3.3.6 and 4.4.2 is how trailing slashes are treated in any source or destination URL. In 3.3.6 all trailing slashes were incorrectly removed making it difficult to specify directories and generally produced strange results as compared to cp or scp. This error was corrected in the 4.x series but, as Gerard noted, produced different results between the two versions. Arguably, the 4.2 xrdcp is correct in the way it handles URL's with a trailing slash (well, except for one bug that will soon be corrected -- when automatic path creation is on and the final directory component does not exist). The 3.x xrdcp is blatantly wrong.

All that said, the question is whether we should allow selecting incorrect behavior for comparability's sake. That is not so clear since it's doubtful people actually specified URL's with a trailing slash using the old xrdcp as it would produce a weird result. I am not totally against it but question whether it is really needed in practice. If we do allow reverting to bad behavior then it would only be via envar (e.g. XRD_STRIPSLASH).

So, now that this is explained, what are people's thoughts on it?.

**Comment by simonmichal** [comment]
I absolutely agree, we should not by default revert to incorrect behaviour because of compatibility reasons.
Providing compatibility through envar should be enough (but only if really necessary).

**Comment by abh3** [comment]
I believe we resolved this as a problem with the particular dCache version being used. Testing revealed that the current backward comparability option is sufficient for operational purposes.

**Comment by gerardba** [comment]
the actual solution was using 'xrootd-old' with the particular dCache
version (dcache-2.2.29).

Gerard

2015-10-21 23:28 GMT-05:00 Andrew Hanushevsky <notifications@github.com>:

**Comment by xrootd-dev** [comment]
Hi Gerard,

That’s unfortunate. But my testing showed this is very specific to the dCache version being used. The 4.x behavior with the envar set to make directories works fine for other storage systems. I have no way to see what it does with the latest version of dCache.

Andy
